src/ring_theory/zmod.lean

```diff
@@ -0,0 +1,36 @@
+/-
+Copyright (c) 2022 Alex J. Best. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex J. Best
+-/
+import algebra.squarefree
+import data.zmod.basic
+import ring_theory.int.basic
+
+/-!
+# Ring theoretic facts about `zmod n`
+
+## Main statements
+
+* `zmod.is_reduced` - `zmod n` is reduced for all squarefree `n`.
+
+-/
+
+instance {n : ℕ} [fact $ squarefree n] : is_reduced (zmod n) :=
```

Comment:
Why do you use `fact`? Did you see somewhere the combination of `fact` with `squarefree`? Is there some application that requires you make it an instance?

Reply:
I have a project with lots of `squarefree n` facts, for mathlib its unclear to me what the best way of setting things up is, either the user has to do
```
haveI : fact (squarefree n) := blah,
apply some_lemma_that_needs_is_reduced
```
or
```
haveI : is_reduced (zmod n) := zmod.is_reduced blah,
apply some_lemma_that_needs_is_reduced
```
so neither is perfect, but at least in the former case there may be other uses for the `fact` (but not in mathlib right now ;))

Reply:
I'd like to see a version with explicit argument `(h : squarefree n)` (maybe `n` explicit as well) because I imagine users who apply this to a concrete type may want to do `haveI : zmod.is_reduced n (by norm_num)`. I'm not sure about the general rules regarding accepting `fact` instances into mathlib; my impression is that `fact` tends to slow down things because it's not a dedicated typeclass. I would recommend that you restrict its scope as much as possible, maybe only make it an instance only in the file you need it. (i.e. I think it's OK to add `def {n : ℕ} [fact $ squarefree n] : is_reduced (zmod n) :=` here but only add `local attribute [instance] zmod.is_reduced` in your files).

Reply:
Well let me start by saying I share your dislike of `fact` out of principle, but I don't think we need to worry about slowdown due to the use of fact here, this isn't an `instance : fact blah` but rather an instance taking `fact` as a parameter. Such lookups can be slow but the only way this would occur is if the user needs `is_reduced (zmod n)` for something, this wouldn't slow down the rest of mathlib any more than adding another normal typeclass instance would as far as I can tell.

Reply:
Thanks for the explanation. I agree that this instance alone won't cause problems. It won't be useful unless there are instances around with `fact (squarefree n)` as conclusion, and I those do slow things down, though. However, we may make them lemmas and only mark them instances locally when needed; is that the plan?

I'm sorry my PR is going slowly because we realized the equivalence can be generalized to GCD domains, but it would depend on [Cauchy induction](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Cauchy.20Induction/near/304871110) (now at #15880, but I'll probably open another PR with the generalization suggested by @vihdzp). I'm also adding definitions for radical elements and radical ideals along with a small amount of API.